api/view_only_links/serializers.py

```diff
@@ -0,0 +1,154 @@
+from rest_framework import serializers as ser
+from rest_framework.exceptions import PermissionDenied, NotFound, ValidationError
+
+from api.base.exceptions import RelationshipPostMakesNoChanges
+from api.base.serializers import (
+    JSONAPISerializer, IDField, RelationshipField,
+    JSONAPIRelationshipSerializer, LinksField, relationship_diff
+)
+from api.base.utils import absolute_reverse
+
+from website.project.model import Node
+
+
+class ViewOnlyLinkDetailSerializer(JSONAPISerializer):
+    key = ser.CharField(read_only=True)
+    id = IDField(source='_id', read_only=True)
+    date_created = ser.DateTimeField(read_only=True)
+    anonymous = ser.BooleanField(required=False)
+    name = ser.CharField(required=False)
+
+    nodes = RelationshipField(
+        related_view='view-only-links:view-only-link-nodes',
+        related_view_kwargs={'link_id': '<_id>'},
+        self_view='view-only-links:view-only-link-nodes-relationships',
+        self_view_kwargs={'link_id': '<_id>'}
+    )
+
+    creator = RelationshipField(
+        related_view='users:user-detail',
+        related_view_kwargs={'user_id': '<creator._id>'},
+    )
+
+    def get_absolute_url(self, obj):
+        return absolute_reverse(
+            'nodes:node-view-only-link-detail',
+            kwargs={
+                'link_id': obj._id
+            }
+        )
+
+    class Meta:
+        type_ = 'view-only-links'
+
+
+class VOLNode(JSONAPIRelationshipSerializer):
+    id = ser.CharField(source='_id')
+
+    class Meta:
+        type_ = 'nodes'
+
+
+class ViewOnlyLinkNodesSerializer(ser.Serializer):
+    data = ser.ListField(child=VOLNode())
+    links = LinksField({
+        'self': 'get_self_url',
+    })
+
+    def get_self_url(self, obj):
+        return absolute_reverse(
+            'view-only-links:view-only-link-nodes',
+            kwargs={
+                'link_id': obj['self']._id
+            }
+        )
+
+    def make_instance_obj(self, obj):
+        return {
+            'data': obj.nodes,
+            'self': obj
+        }
+
+    def get_nodes_to_add_remove(self, nodes, new_nodes):
+        diff = relationship_diff(
+            current_items={node._id: node for node in nodes},
+            new_items={node['_id']: node for node in new_nodes}
+        )
+
+        nodes_to_add = []
+        for node_id in diff['add']:
+            node = Node.load(node_id)
+            if not node:
+                raise NotFound
+            nodes_to_add.append(node)
+
+        return nodes_to_add, diff['remove'].values()
+
+    def get_eligible_nodes(self, nodes):
+        return [
+            descendant
+            for node in nodes
+            for descendant in node.get_descendants_recursive()
```

Comment:
The function needs to somehow iterate over all the nodes currently associated with the VOL. Since view only links don't have a concept of a primary or parent node, checking the descendants of a single node in the list of nodes doesn't guarantee that we're actually checking all of the eligible descendants. 

Additionally, ```node.get_descendants_recursive()``` returns a generator that needs to be iterated over to add all of the descendants to the eligible nodes list.

It's a bit complicated, but the nested loops are necessary :/